lib/puppet/configurer.rb

```diff
@@ -393,32 +389,20 @@ def run_internal(options)
   end
   private :run_internal
 
-  def find_functional_server()
-    configured_environment = Puppet[:environment] if Puppet.settings.set_by_config?(:environment)
-
-    node = nil
-    selected_server = Puppet.settings[:server_list].find do |server|
-      # Puppet.override doesn't return the result of its block, so we
-      # need to handle this manually
-      found = false
-      server[1] ||= Puppet[:masterport]
-      Puppet.override(:server => server[0], :serverport => server[1]) do
-        begin
-          node = Puppet::Node.indirection.find(Puppet[:node_name_value],
-              :environment => Puppet::Node::Environment.remote(@environment),
-              :configured_environment => configured_environment,
-              :ignore_cache => true,
-              :transaction_uuid => @transaction_uuid,
-              :fail_on_404 => false)
-          found = true
-        rescue
-          # Nothing to see here
-        end
+  def find_functional_server
+    Puppet.settings[:server_list].each do |server|
+      host = server[0]
+      port = server[1] || Puppet[:masterport]
+      begin
+        http = Puppet::Network::HttpPool.http_ssl_instance(host, port)
+        http.head('/status/v1/simple')
+        return [host, port]
```

Comment:
This looks like the request is considered successful if it gets an HTTP response at all? Even if it's a `404`/`500`? What's the reason for not making sure the request was a `200`?

Reply:
Yeah, we should make sure the request was successful. It turns out that calling HEAD on the status endpoint returns 404, but `GET` works:

```
$ curl -v -k -I https://localhost:8140/status/v1/simple
*   Trying ::1...
* Connected to localhost (::1) port 8140 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
* Server certificate: localhost
> HEAD /status/v1/simple HTTP/1.1
> Host: localhost:8140
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 404 Not Found
HTTP/1.1 404 Not Found
```